### PR TITLE
Keep a queue of sensors to connect to

### PR DESCRIPTION
**Pull request by qwandor-google**
This series of changes implements a queue of sensors which we want to connect to, with a configurable size limit, and named sensors having priority. The main loop will keep trying to connect to them, interleaved with receiving events from connected sensors.

### CODE REVIEW

File: mijia/src/lib.rs [hunk 1]
```diff
@@ -76,13 +76,20 @@ pub fn print_sensors(sensors: &[BluetoothDevice], sensor_names: &HashMap<String,
     }
 }
 
+pub fn connect_sensor<'a>(sensor: &BluetoothDevice<'a>) -> bool {
+    if let Err(e) = sensor.connect(10000) {
```
Comment (alsuren) [inline mijia/src/lib.rs:80]:
```suggestion
    if let Err(e) = sensor.connect(10_000) {
```

File: publish-mqtt/src/main.rs [hunk 2]
```diff
@@ -106,55 +110,110 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
...
+                &mut homie,
+                &sensor_names,
+                properties.to_vec(),
+                &sensor,
+            )
+            .await
+            {
+                println!("Failed to connect to {}: {:?}", name, e);
+                sensors_to_connect.push_back(sensor);
+            } else {
```
Comment (alsuren) [inline publish-mqtt/src/main.rs:201]: (optional) `if let` with an `else` block feels a bit odd. I think I would use match here.

File: publish-mqtt/src/main.rs [hunk 3]
```diff
@@ -106,55 +110,110 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
...
     }
-    homie.publish_nodes().await?;
 
-    start_notify_sensors(bt_session, &connected_sensors);
+    homie.ready().await?;
 
     // currently there is no way to set INCOMING_TIMEOUT_MS to -1 (which is how
     // you specify "infinite" according to
     // https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#ga580d8766c23fe5f49418bc7d87b67dc6)
     // so we wrap the event loop in an infinite loop.
```
Comment (alsuren) [inline publish-mqtt/src/main.rs:182]: this comment is now outdated

File: publish-mqtt/src/main.rs [hunk 4]
```diff
@@ -106,55 +110,110 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
...
-        let node_name = sensor_names.get(&mac_address).unwrap_or(&mac_address);
-        homie.add_node(Node::new(
-            node_id,
-            node_name.clone(),
-            "Mijia sensor".to_string(),
-            properties.clone(),
-        ));
+
+    let mut sensors_to_connect: VecDeque<_> = named_sensors.into();
+    if sensors_to_connect.len() < max_connected_sensors {
```
Comment (alsuren) [inline publish-mqtt/src/main.rs:165]: at some point, we will need to think about the case where named_sensors.len() > max_connected_sensors. I have made a card for it.

**Review by alsuren — APPROVED** [review]
Looks good. Entirely optional style comments.

File: mijia/src/lib.rs (hunk 1 above)
Comment (qwandor-google) [reply mijia/src/lib.rs:80]: Changed to a constant.

File: publish-mqtt/src/main.rs (hunk 2 above)
Comment (qwandor-google) [reply publish-mqtt/src/main.rs:201]: Done.

File: publish-mqtt/src/main.rs (hunk 3 above)
Comment (qwandor-google) [reply publish-mqtt/src/main.rs:182]: Removed.

**Pull request merged**